parsons/databases/postgres/postgres_create_statement.py

```diff
@@ -0,0 +1,242 @@
+
+        if varchar_max:
+            mapping['longest'] = self.vc_max(mapping, varchar_max)
+
+        if varchar_truncate:
+            mapping['longest'] = self.vc_trunc(mapping)
+
+        mapping['longest'] = self.vc_validate(mapping)
+
+        # Add any provided column type overrides
+        if columntypes:
+            for i in range(len(mapping['headers'])):
+                col = mapping['headers'][i]
+                if columntypes.get(col):
+                    mapping['type_list'][i] = columntypes[col]
+
+        # Enclose in quotes
+        mapping['headers'] = [f'"{h}"'for h in mapping['headers']]
+
+        return self.create_sql(table_name, mapping, distkey=distkey, sortkey=sortkey)
+
+    def data_type(self, val, current_type):
+        # Determine the database data type of a given value
+
+        try:
+            # Convert to string to reevaluate data type
+            t = ast.literal_eval(str(val))
+        except ValueError:
+            return 'varchar'
+        except SyntaxError:
+            return 'varchar'
+
+        if type(t) in [int, float]:
+            if (type(t) in [int] and current_type not in ['float', 'varchar']):
+
+                # Make sure that it is a valid integer
+                if not self.is_valid_integer(val):
+                    return 'varchar'
+
+                # Use smallest possible int type
+                if (-32768 < t < 32767) and current_type not in ['int', 'bigint']:
+                    return 'smallint'
+                elif (-2147483648 < t < 2147483647) and current_type not in ['bigint']:
+                    return 'int'
+                else:
+                    return 'bigint'
+            if type(t) is float and current_type not in ['varchar']:
+                return 'decimal'
+        else:
+            return 'varchar'
+
+    def is_valid_integer(self, val):
+
+        # Valid ints in python can contain an underscore, but Postgres can't. This
+        # checks to see if there is an underscore in the value and turns it into
+        # a varchar if so.
+        try:
+            if '_' in val:
+                return False
+
+        except TypeError:
+            return True
+
+        # If it has a leading zero, we should treat it as a varchar, since it is
+        # probably there for a good reason (e.g. zipcode)
+        if val.isdigit():
+            if val[0] == '0':
+                return False
+
+        return True
+
+    def generate_data_types(self, table):
+        # Generate column data types
+
+        longest, type_list = [], []
+
+        cont = petl.records(table.table)
+
+        # Populate empty values for the columns
+        for col in table.columns:
+            longest.append(0)
+            type_list.append('')
+
+        for row in cont:
+            for i in range(len(row)):
+                # NA is the csv null value
+                if type_list[i] == 'varchar' or row[i] == 'NA':
+                    pass
```

Comment:
If we `pass` here, does that mean we aren't going to count the characters in this string to increase the size of our `longest`? You could make the `else` below an `elif` and check whether it's `varchar` there so.

Reply:
You're right. Sorry. I read the `pass` as `continue`.